File: core/src/androidTest/java/com/emarsys/core/DeviceInfoTest.java
package com.emarsys.core;

import android.app.Application;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.content.res.Resources;

import com.emarsys.core.device.DeviceInfo;
import com.emarsys.core.device.LanguageProvider;
import com.emarsys.core.notification.NotificationManagerHelper;
import com.emarsys.core.provider.hardwareid.HardwareIdProvider;
import com.emarsys.core.provider.version.VersionProvider;
import com.emarsys.testUtil.ApplicationTestUtils;
import com.emarsys.testUtil.InstrumentationRegistry;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import java.util.Locale;
import java.util.TimeZone;

import static junit.framework.Assert.assertFalse;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class DeviceInfoTest {

    private static final String HARDWARE_ID = "hwid";
    private static final String SDK_VERSION = "sdkVersion";
    private static final String LANGUAGE = "en-US";

    private DeviceInfo deviceInfo;
    private TimeZone tz;
    private Context context;
    private HardwareIdProvider mockHardwareIdProvider;
    private LanguageProvider mockLanguageProvider;
    private VersionProvider mockVersionProvider;
    private NotificationManagerHelper mockNotificationManagerHelper;
    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void setup() {
        tz = TimeZone.getTimeZone("Asia/Tokyo");
        TimeZone.setDefault(tz);

        context = InstrumentationRegistry.getTargetContext().getApplicationContext();

        mockHardwareIdProvider = mock(HardwareIdProvider.class);
        mockLanguageProvider = mock(LanguageProvider.class);
        mockVersionProvider = mock(VersionProvider.class);
        mockNotificationManagerHelper = mock(NotificationManagerHelper.class);

        when(mockHardwareIdProvider.provideHardwareId()).thenReturn(HARDWARE_ID);
        when(mockLanguageProvider.provideLanguage(any(Locale.class))).thenReturn(LANGUAGE);
        when(mockVersionProvider.provideSdkVersion()).thenReturn(SDK_VERSION);

        deviceInfo = new DeviceInfo(context, mockHardwareIdProvider, mockVersionProvider, mockLanguageProvider, mockNotificationManagerHelper, true);
    }

    @After
    public void teardown() {
        TimeZone.setDefault(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_context_mustNotBeNull() {
        new DeviceInfo(null, mockHardwareIdProvider, mockVersionProvider, mockLanguageProvider, mockNotificationManagerHelper, true);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_hardwareIdProvider_mustNotBeNull() {
        new DeviceInfo(context, null, mockVersionProvider, mockLanguageProvider, mockNotificationManagerHelper, true);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_versionProvider_mustNotBeNull() {
        new DeviceInfo(context, mockHardwareIdProvider, null, mockLanguageProvider, mockNotificationManagerHelper, true);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_languageProvider_mustNotBeNull() {
        new DeviceInfo(context, mockHardwareIdProvider, mockVersionProvider, null, mockNotificationManagerHelper, true);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_notificationManagerHelper_mustNotBeNull() {
        new DeviceInfo(context, mockHardwareIdProvider, mockVersionProvider, mockLanguageProvider, null, true);
    }

    @Test
    public void testConstructor_initializesFields() {
        assertNotNull(deviceInfo.getHwid());
        assertNotNull(deviceInfo.getPlatform());
        assertNotNull(deviceInfo.getLanguage());
        assertNotNull(deviceInfo.getTimezone());
        assertNotNull(deviceInfo.getManufacturer());
        assertNotNull(deviceInfo.getModel());
        assertNotNull(deviceInfo.getApplicationVersion());
        assertNotNull(deviceInfo.getOsVersion());
        assertNotNull(deviceInfo.getDisplayMetrics());
        assertNotNull(deviceInfo.getSdkVersion());
    }

    @Test
    public void testGetApplicationVersion_shouldBeDefault_whenVersionInPackageInfo_isNull() throws PackageManager.NameNotFoundException {
        String packageName = "packageName";
        Context mockContext = mock(Context.class);
        PackageInfo packageInfo = new PackageInfo();
        PackageManager packageManager = mock(PackageManager.class);
        packageInfo.versionName = null;

        when(mockContext.getContentResolver()).thenReturn(InstrumentationRegistry.getTargetContext().getContentResolver());
        when(mockContext.getPackageName()).thenReturn(packageName);
        when(mockContext.getPackageManager()).thenReturn(packageManager);
        when(packageManager.getPackageInfo(packageName, 0)).thenReturn(packageInfo);
        when(mockContext.getApplicationInfo()).thenReturn(mock(ApplicationInfo.class));

        DeviceInfo info = new DeviceInfo(mockContext, mockHardwareIdProvider, mockVersionProvider, mockLanguageProvider, mockNotificationManagerHelper, true);

        assertEquals(DeviceInfo.UNKNOWN_VERSION_NAME, info.getApplicationVersion());
    }

    @Test
    public void testTimezoneCorrectlyFormatted() {
        assertEquals("+0900", deviceInfo.getTimezone());
    }

    @Test
    public void testTimezoneCorrectlyFormatted_withArabicLocale() {
        Locale previous = Locale.getDefault();

        Locale locale = new Locale("my");
        Resources resources = InstrumentationRegistry.getTargetContext().getResources();
        Locale.setDefault(locale);
        Configuration config = resources.getConfiguration();
        config.locale = locale;
        resources.updateConfiguration(config, resources.getDisplayMetrics());

        Locale.setDefault(previous);

        assertEquals("+0900", deviceInfo.getTimezone());
    }

    @Test
    public void testGetDisplayMetrics() {
        assertEquals(deviceInfo.getDisplayMetrics(), Resources.getSystem().getDisplayMetrics());
    }

    @Test
    public void testIsDebugMode_withDebugApplication() {
        Application mockDebugContext = ApplicationTestUtils.getApplicationDebug();

        DeviceInfo debugDeviceInfo = new DeviceInfo(mockDebugContext, mockHardwareIdProvider, mockVersionProvider, mockLanguageProvider, mockNotificationManagerHelper, true);
        assertTrue(debugDeviceInfo.isDebugMode());
    }

    @Test
    public void testIsDebugMode_withReleaseApplication() {
        Application mockReleaseContext = ApplicationTestUtils.getApplicationRelease();

        DeviceInfo releaseDeviceInfo = new DeviceInfo(mockReleaseContext, mockHardwareIdProvider, mockVersionProvider, mockLanguageProvider, mockNotificationManagerHelper, true);
        assertFalse(releaseDeviceInfo.isDebugMode());
    }

    @Test
    public void testHardwareId_isAcquiredFromHardwareIdProvider() {
        verify(mockHardwareIdProvider).provideHardwareId();
        assertEquals(HARDWARE_ID, deviceInfo.getHwid());
    }

    @Test
    public void testGetLanguage_isAcquiredFromLanguageProvider() {
        String language = deviceInfo.getLanguage();

        verify(mockLanguageProvider).provideLanguage(Locale.getDefault());

        assertEquals(LANGUAGE, language);
    }

    @Test
    public void testGetHash_shouldEqualHashCode() {
        Integer expectedHash = deviceInfo.hashCode();

        assertEquals(expectedHash, deviceInfo.getHash());
    }

    @Test
    public void testGetNotificationSettings() {
        assertEquals(deviceInfo.getNotificationSettings(), mockNotificationManagerHelper);
    }
}

File: core/src/androidTest/java/com/emarsys/core/DeviceInfoTest.kt
package com.emarsys.core

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.res.Resources
import android.os.Build
import com.emarsys.core.api.notification.ChannelSettings
import com.emarsys.core.api.notification.NotificationSettings
import com.emarsys.core.device.DeviceInfo
import com.emarsys.core.device.LanguageProvider
import com.emarsys.core.provider.clientid.ClientIdProvider
import com.emarsys.core.provider.version.VersionProvider
import com.emarsys.core.util.AndroidVersionUtils
import com.emarsys.testUtil.ApplicationTestUtils.applicationDebug
import com.emarsys.testUtil.ApplicationTestUtils.applicationRelease
import com.emarsys.testUtil.InstrumentationRegistry.Companion.getTargetContext
import io.kotest.matchers.booleans.shouldBeFalse
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.json.JSONObject
import org.junit.After
import org.junit.Before
import org.junit.Test
import java.util.*

class DeviceInfoTest {

    companion object {
        private const val CLIENT_ID = "hwid"
        private const val SDK_VERSION = "sdkVersion"
        private const val LANGUAGE = "en-US"
        private const val APP_VERSION = "2.0"
        private const val PACKAGE_NAME = "packageName"
    }

    private lateinit var deviceInfo: DeviceInfo
    private lateinit var tz: TimeZone
    private lateinit var context: Context
    private lateinit var mockClientIdProvider: ClientIdProvider
    private lateinit var mockLanguageProvider: LanguageProvider
    private lateinit var mockVersionProvider: VersionProvider
    private lateinit var mockNotificationManagerHelper: NotificationSettings
    private lateinit var mockContext: Context
    private lateinit var mockPackageManager: PackageManager


    @Before
    fun setup() {
        tz = TimeZone.getTimeZone("Asia/Tokyo")
        TimeZone.setDefault(tz)
        context = getTargetContext().applicationContext
        mockClientIdProvider = mockk(relaxed = true)
        every { mockClientIdProvider.provideClientId() } returns CLIENT_ID

        mockLanguageProvider = mockk(relaxed = true)
        every { mockLanguageProvider.provideLanguage(any()) } returns LANGUAGE

        mockVersionProvider = mockk(relaxed = true)
        every { mockVersionProvider.provideSdkVersion() } returns SDK_VERSION

        mockNotificationManagerHelper = mockk(relaxed = true)
        mockContext = mockk(relaxed = true)
        mockPackageManager = mockk(relaxed = true)

        every { mockContext.contentResolver } returns getTargetContext().contentResolver
        every { mockContext.packageName } returns PACKAGE_NAME
        every { mockContext.packageManager } returns mockPackageManager

        deviceInfo = DeviceInfo(
            context, mockClientIdProvider, mockVersionProvider,
            mockLanguageProvider, mockNotificationManagerHelper,
            isAutomaticPushSendingEnabled = true, isGooglePlayAvailable = true
        )
    }

    @After
    fun teardown() {
        TimeZone.setDefault(null)
    }

    @Test
    fun testConstructor_initializesFields() {
        with(deviceInfo) {
            clientId shouldNotBe null
            platform shouldNotBe null
            language shouldNotBe null
            timezone shouldNotBe null
            manufacturer shouldNotBe null
            model shouldNotBe null
            applicationVersion shouldNotBe null
            osVersion shouldNotBe null
            displayMetrics shouldNotBe null
            sdkVersion shouldNotBe null
            notificationSettings shouldNotBe null
        }
    }

    @Test
    @Throws(PackageManager.NameNotFoundException::class)
    fun testGetApplicationVersion_shouldBeDefault_whenVersionInPackageInfo_isNull() {
        val packageInfo = PackageInfo()
        packageInfo.versionName = null

        every { mockContext.applicationInfo } returns mockk(relaxed = true)
        val info = DeviceInfo(
            mockContext, mockClientIdProvider, mockVersionProvider,
            mockLanguageProvider, mockNotificationManagerHelper,
            isAutomaticPushSendingEnabled = true, isGooglePlayAvailable = true
        )
        info.applicationVersion shouldBe DeviceInfo.UNKNOWN_VERSION_NAME
    }

    @Test
    fun testTimezoneCorrectlyFormatted() {
        "+0900" shouldBe deviceInfo.timezone
    }

    @Test
    fun testTimezoneCorrectlyFormatted_withArabicLocale() {
        val previous = Locale.getDefault()
        val locale = Locale("my")
        val resources = getTargetContext().resources
        Locale.setDefault(locale)
        val config = resources.configuration
        config.setLocale(locale)
        context.createConfigurationContext(config)
        Locale.setDefault(previous)
        "+0900" shouldBe deviceInfo.timezone
    }

    @Test
    fun testGetDisplayMetrics() {
        deviceInfo.displayMetrics shouldBe Resources.getSystem().displayMetrics
    }

    @Test
    fun testIsDebugMode_withDebugApplication() {
        val mockDebugContext = applicationDebug
        val debugDeviceInfo = DeviceInfo(
            mockDebugContext, mockClientIdProvider,
            mockVersionProvider, mockLanguageProvider, mockNotificationManagerHelper,
            isAutomaticPushSendingEnabled = true, isGooglePlayAvailable = true
        )
        debugDeviceInfo.isDebugMode.shouldBeTrue()
    }

    @Test
    fun testIsDebugMode_withReleaseApplication() {
        val mockReleaseContext = applicationRelease
        val releaseDeviceInfo = DeviceInfo(
            mockReleaseContext,
            mockClientIdProvider,
            mockVersionProvider,
            mockLanguageProvider,
            mockNotificationManagerHelper,
            isAutomaticPushSendingEnabled = true,
            isGooglePlayAvailable = true
        )
        releaseDeviceInfo.isDebugMode.shouldBeFalse()
    }

    @Test
    fun testClientId_isAcquiredFromClientIdProvider() {
        verify { mockClientIdProvider.provideClientId() }
        CLIENT_ID shouldBe deviceInfo.clientId
    }

    @Test
    fun testGetLanguage_isAcquiredFromLanguageProvider() {
        val language = deviceInfo.language
        verify { mockLanguageProvider.provideLanguage(Locale.getDefault()) }
        LANGUAGE shouldBe language
    }

    @Test
    fun testGetDeviceInfoPayload_shouldEqualPayload() {
        val packageInfo = PackageInfo()
        packageInfo.versionName = APP_VERSION
        every { mockPackageManager.getPackageInfo(PACKAGE_NAME, 0) } returns packageInfo

        deviceInfo = DeviceInfo(
            mockContext, mockClientIdProvider, mockVersionProvider,
            mockLanguageProvider, mockNotificationManagerHelper,
            isAutomaticPushSendingEnabled = true, isGooglePlayAvailable = true
        )

        every { mockNotificationManagerHelper.channelSettings } returns listOf(
            ChannelSettings(
                channelId = "channelId"
            )
        )

        var channelSettings = """
        channelSettings: [
            {
                "channelId":"channelId",
                "importance":-1000,
                "isCanBypassDnd":false,
                "isCanShowBadge":false,
                "isShouldVibrate":false
            }
        ]"""

        if (!AndroidVersionUtils.isOreoOrAbove) {
            channelSettings = "channelSettings: [{}]"
        }
        val expectedPayload = JSONObject(
            """{
                  "notificationSettings": {
                    $channelSettings,
                    "importance": 0,
                    "areNotificationsEnabled": false
                  },
                  "hwid": "hwid",
                  "platform": "android",
                  "language": "en-US",
                  "timezone": "+0900",
                  "manufacturer": "${Build.MANUFACTURER}",
                  "model": "${Build.MODEL}",
                  "osVersion": "${Build.VERSION.RELEASE}",
                  "displayMetrics": "${Resources.getSystem().displayMetrics.widthPixels}x${Resources.getSystem().displayMetrics.heightPixels}",
                  "sdkVersion": "sdkVersion",
                  "appVersion": "$APP_VERSION" 
                }"""
        ).toString()
        deviceInfo.deviceInfoPayload shouldBe expectedPayload
    }

    @Test
    fun testDeviceInfo_platformShouldBeHuawei() {
        val packageInfo = PackageInfo()
        packageInfo.versionName = APP_VERSION
        every { mockPackageManager.getPackageInfo(PACKAGE_NAME, 0) } returns packageInfo

        deviceInfo = DeviceInfo(
            mockContext, mockClientIdProvider, mockVersionProvider,
            mockLanguageProvider, mockNotificationManagerHelper,
            isAutomaticPushSendingEnabled = true, isGooglePlayAvailable = false
        )

        deviceInfo.platform shouldBe "android-huawei"
    }

    @Test
    fun testGetNotificationSettings() {
        deviceInfo.notificationSettings shouldBe mockNotificationManagerHelper
    }

}